main/src/ca/uwaterloo/flix/language/phase/Safety.scala

```diff
@@ -33,303 +33,406 @@ object Safety {
+  private def visitSendable(root: Root)(implicit flix: Flix): Validation[List[Instance], CompilationMessage] = {
 
     val sendableClass = new Symbol.ClassSym(Nil, "Sendable", SourceLocation.Unknown)
-
-    root.instances.getOrElse(sendableClass, Nil) flatMap {
-      case Instance(_, _, _, _, tpe, _, _, _, loc) =>
-        if (tpe.typeArguments.exists(_.kind == Kind.Bool))
-          List(SafetyError.SendableError(tpe, loc))
-        else
-          Nil
+    traverse(root.instances.getOrElse(sendableClass, Nil)) {
+      case Instance(doc, ann, mod, clazz, tpe, tconstrs, defs, ns, loc) =>
+        if (tpe.typeArguments.exists(_.kind == Kind.Bool)) {
+          SafetyError.SendableError(tpe, loc).toFailure
+        } else {
+          Instance(doc, ann, mod, clazz, tpe, tconstrs, defs, ns, loc).toSuccess
+        }
     }
   }
 
   /**
     * Performs safety and well-formedness checks on the given definition `def0`.
     */
-  private def visitDef(def0: Def, root: Root)(implicit flix: Flix): List[CompilationMessage] = {
+  private def visitDef(def0: Def, root: Root)(implicit flix: Flix): Validation[Root, CompilationMessage] = {
     val renv = def0.spec.tparams.map(_.sym).foldLeft(RigidityEnv.empty) {
       case (acc, e) => acc.markRigid(e)
     }
-    visitExp(def0.impl.exp, renv, root)
+    mapN(visitExp(def0.impl.exp, renv, root))(_ => root)
   }
 
 
   /**
     * Performs safety and well-formedness checks on the given expression `exp0`.
     */
-  private def visitExp(e0: Expression, renv: RigidityEnv, root: Root)(implicit flix: Flix): List[CompilationMessage] = {
+  private def visitExp(e0: Expression, renv: RigidityEnv, root: Root)(implicit flix: Flix): Validation[Expression, CompilationMessage] = {
 
-    def visit(exp0: Expression): List[CompilationMessage] = exp0 match {
-      case Expression.Cst(_, _, _) => Nil
+    def visit(exp0: Expression): Validation[Expression, CompilationMessage] = exp0 match {
+      case Expression.Cst(cst, tpe, loc) => Expression.Cst(cst, tpe, loc).toSuccess
 
-      case Expression.Wild(_, _) => Nil
+      case Expression.Wild(tpe, loc) => Expression.Wild(tpe, loc).toSuccess
 
-      case Expression.Var(_, _, _) => Nil
+      case Expression.Var(sym, tpe, loc) => Expression.Var(sym, tpe, loc).toSuccess
 
-      case Expression.Def(_, _, _) => Nil
+      case Expression.Def(sym, tpe, loc) => Expression.Def(sym, tpe, loc).toSuccess
 
-      case Expression.Sig(_, _, _) => Nil
+      case Expression.Sig(sym, tpe, loc) => Expression.Sig(sym, tpe, loc).toSuccess
 
-      case Expression.Hole(_, _, _) => Nil
+      case Expression.Hole(sym, tpe, loc) => Expression.Hole(sym, tpe, loc).toSuccess
 
-      case Expression.HoleWithExp(exp, _, _, _, _) =>
-        visit(exp)
+      case Expression.HoleWithExp(exp, tpe, pur, eff, loc) =>
+        val expVal = visit(exp)
+        mapN(expVal) { e => Expression.HoleWithExp(e, tpe, pur, eff, loc) }
```

Comment:
Style note: We almost never have lambdas on the same line as `map`. Usually it looks like

```scala
mapN(expVal) { 
    e => Expression.HoleWithExp(e, tpe, pur, eff, loc)
}
```

Reply:
Will keep that in mind going forward.